lib/neverblock/core/fiber.rb

```diff
@@ -35,10 +35,88 @@ def self.yield(*args)
   end
 
   private
-  
+
   def local_fiber_variables
     @local_fiber_variables ||= {}
   end
 
 end
 
+# Code that is designed to work with mutexes and threads but isn't aware of fibers
+# will potentially screw up. 
+class Mutex
+  alias :_orig_try_lock :try_lock
+  def try_lock
+    acquired = _orig_try_lock
+    disable_neverblock if acquired
+    acquired
+  end
+
+  # sleep is mostly used by ConditionVariable
+  alias :_orig_sleep :sleep
+  def sleep(*args)
+    restore_neverblock
+    ret = _orig_sleep(*args)
+    disable_neverblock
+    ret
+  end
+
+  alias :_orig_lock :lock
+  def lock
```

Comment:
given lock A, lock B. 
(1) lock A
(2) lock B
(3) unlock A
=> Neverblocking == true # which it shouldnt be yet
(4) unlock B
=> Neverblocking == false # which at this point is should be true

Reply:
Yes that would mess up.  The more common case of lock A, lock B, unlock B, unlock A does work though. I would guess no code would do this but it'd suck cause it'd turn off neverblock for a bit. So the :neverblock variable is set again for every fiberpool.spawn which is what the amqp.handle_message! dispatch function uses so it'd at least get reset once that fiber was re-used at least. Will try and think of a fix that's simple enough to think through...

Reply:
Alright -- redid the disable/restore functions. No instance variable, keeps tracks of things per-fiber. For every AMWP request, fiber.spawn is called which re-uses a fiber from the pool (passing a block to its resume function) but resets the fiber.neverblock and fiber.lock_count variables which should mean something unexpected shouldn't screw up later work/other fibers.

Reply:
Also added a spec for this case